cpp/src/arrow/compute/kernels/vector_dictionary.cc

```diff
@@ -0,0 +1,210 @@
+
+namespace compute {
+namespace internal {
+
+namespace {
+
+// Dictionary compaction implementation
+
+const FunctionDoc dictionary_compact_doc{
+    "Compact dictionary array",
+    ("Return a compacted version of the dictionary array input,\n"
+     "which removes unused values in dictionary.\n"
+     "The function assumes every indice is effective."),
+    {"dictionary_array"}};
+
+class DictionaryCompactKernel : public KernelState {
+ public:
+  virtual Result<std::shared_ptr<Array>> Exec(std::shared_ptr<Array> dict_array,
+                                              ExecContext* ctx) const = 0;
+};
+
+template <typename IndiceArrowType>
+class DictionaryCompactKernelImpl : public DictionaryCompactKernel {
+  using BuilderType = NumericBuilder<IndiceArrowType>;
+  using CType = typename IndiceArrowType::c_type;
+
+ public:
+  Result<std::shared_ptr<Array>> Exec(std::shared_ptr<Array> dict_array,
+                                      ExecContext* ctx) const override {
+    const DictionaryArray& casted_dict_array =
+        checked_cast<const DictionaryArray&>(*dict_array);
+    const std::shared_ptr<Array>& dict = casted_dict_array.dictionary();
+    if (dict->length() == 0) {
+      return dict_array;
+    }
+    const std::shared_ptr<Array>& indice = casted_dict_array.indices();
+    if (indice->length() == 0) {
+      ARROW_ASSIGN_OR_RAISE(auto empty_dict,
+                            MakeEmptyArray(dict->type(), ctx->memory_pool()));
+      return DictionaryArray::FromArrays(dict_array->type(), indice, empty_dict);
+    }
+    const CType* indices_data =
+        reinterpret_cast<const CType*>(indice->data()->buffers[1]->data());
+    int64_t offset = indice->data()->offset;
+
+    // check whether the input is compacted
+    std::vector<bool> dict_used(dict->length(), false);
+    int64_t dict_used_count = 0;
+    for (int64_t i = 0; i < indice->length(); i++) {
+      if (indice->IsNull(i)) {
+        continue;
+      }
+
+      CType cur_indice = indices_data[i + offset];
+      if (!dict_used[cur_indice]) {
+        dict_used[cur_indice] = true;
+        dict_used_count++;
+
+        if (dict_used_count == dict->length()) {  // input is already compacted
+          return dict_array;
+        }
+      }
+    }
+
+    // dictionary compaction
+    if (dict_used_count == 0) {
+      ARROW_ASSIGN_OR_RAISE(auto empty_dict,
+                            MakeEmptyArray(dict->type(), ctx->memory_pool()));
+      return DictionaryArray::FromArrays(dict_array->type(), indice, empty_dict);
+    }
+    std::vector<CType> dict_indice;
+    bool need_change_indice = false;
+    CType len = static_cast<CType>(dict->length());
+    for (CType i = 0; i < len; i++) {
+      if (dict_used[i]) {
+        dict_indice.push_back(i);
+      } else if (i + 1 < len && dict_used[i + 1]) {
+        need_change_indice = true;
+      }
+    }
+    BuilderType dict_indice_builder;
+    ARROW_RETURN_NOT_OK(dict_indice_builder.AppendValues(dict_indice));
```

Comment:
Wouldn't it be more straightforward to append to the builder in the loop instead of building a vector first?

Reply:
Sure, let me change the code